### PR TITLE
LB-322: Add import and sign in buttons to error pages

### PR DESCRIPTION
**Pull request by Uditgulati**
# Summary

* This is a…
    * [ ] Bug fix
    * [ ] Feature addition
    * [ ] Refactoring
    * [x] Minor / simple change (like a typo)
    * [ ] Other

# Problem

* JIRA ticket (_optional_): [LB-322](https://tickets.metabrainz.org/browse/LB-322)

Sign in and import buttons are not shown on error pages.

# Solution

If it is an error page and user is not signed in, then show import and sign in buttons on that page.

### CODE REVIEW

**Comment by paramsingh** [comment]
Please write a test for this too?

**Comment by Uditgulati** [comment]
Currently, drop-down menu isn't shown on error pages if user is signed in. But, we can show it on 40x pages and show only 'sign in' button or nothing on 500 page. Showing nothing on 500 page is a better choice IMHO.

**Comment by Uditgulati** [comment]
@paramsingh i''ll add a commit for tests once the changes are finalized.

**Comment by paramsingh** [comment]
> Currently, drop-down menu isn't shown on error pages if user is signed in. But, we can show it on 40x pages and show only 'sign in' button or nothing on 500 page. Showing nothing on 500 page is a better choice IMHO.

This is what we're doing in AB, should be okay here too.

**Comment by Uditgulati** [comment]
Now, if error code is not 500, buttons are unchanged. If code is 500, then ``sign in`` button is shown if not signed in, else it shows nothing.

File: listenbrainz/webserver/templates/navbar.html [hunk 1]
```diff
@@ -46,6 +50,16 @@
             </li>
           {% endif %}
         </ul>
+      {% else %}
```
Comment (mayhem) [inline listenbrainz/webserver/templates/navbar.html]: I'm not really a fan of this block of HTML being duplicated. Can you re-work the IF clause to reduce the duplication?

**Comment by Uditgulati** [comment]
Oh, oops. Actually, it was ``500`` instead of ``404``. In ``500``, the page isn't able to render if drop-down menu is shown. 
While cross-checking the if claus, I replaced ``500`` with ``404`` and forgot to revert it back before pushing. I'll fix it.

**Comment by alastair** [comment]
The core idea here is that we can't do any database queries when showing an error page that may have come about due to an error with the database. We require a db query to get information about the current logged in user and show their name in the top menu if they are signed in. This restriction generally applies to 500 errors. We had a case in AB where we saw this causing issues: https://tickets.metabrainz.org/projects/AB/issues/AB-74

We have a similar PR open on AB (https://github.com/metabrainz/acousticbrainz-server/pull/258), and we changed ideas a bit, arriving to the decision that we could probably do this DB query (and therefore show the user menu) on error pages that aren't 500. I think that this PR is incorrect (just by looking at the code, not running it) - we should show the menu if the error is not *500*, not 404. 
Having said that, we should move logic out of the template. I'd prefer to have a template variable `show_user_menu` or similar, which the error handler for 500 can set to False (perhaps we have to invert this flag so that the error handler sets it to True, and therefore we don't have to set it when showing every other page on the site)

**Comment by alastair** [comment]
> While cross-checking the if claus, I replaced 500 with 404 and forgot to revert it back before pushing. I'll fix it.

This is a great indication that we should have a test for this!

**Comment by mayhem** [comment]
Thanks @alastair and @Uditgulati for the explanations -- now it makes sense. 404 threw me off. :)

**Comment by Uditgulati** [comment]
@alastair  I am getting no clue on how to test this as I can't think of a way to create conditions for 500 page. Any ideas?

**Comment by paramsingh** [comment]
@Uditgulati first thing that comes to my mind is that you could add a view to `self.app` in the test that raises InternalServerError and then call that view?

**Comment by Uditgulati** [comment]
@paramsingh I need to check that import button and drop-down menu aren't shown if user is logged in and there is 500 error. Would I need to login somehow in test or is there a way to mock login?

**Comment by paramsingh** [comment]
There is self.temporary_login

**Comment by alastair** [comment]
hi @Uditgulati!
I wrote some tests to check the user menu in the following cases:
 * user logged out
 * user logged in
 * http400 or http404 (show the menu)
 * http500 (don't show the menu)

I also changed the conditions in which we show the menu a bit, now we always show the 'Import now!' button, and then hide the user menu in a special case (`hide_navbar_user_menu`) so that we don't have to have http status codes in the template. Let me know what you think

**Comment by alastair** [comment]
there's a failing test that I missed, I'll look at it later/tomorrow

**Comment by Uditgulati** [comment]
@alastair  I think maybe in http500, we should only show 'Import!' button coz it could be that before this page, user might be logged in or not, so if he is, showing 'Sign In' button would be a bit weird for UI. Btw, thanks a lot for writing tests :)
If u don't mind, I'd like to fix that failing test :)

**Comment by alastair** [comment]
tests working!

**Comment by alastair** [comment]
@mayhem @paramsingh this is ready for merge

**Pull request merged**